pinot-segment-local/src/main/java/org/apache/pinot/segment/local/realtime/converter/stats/MutableColumnStatistics.java

```diff
@@ -70,46 +76,57 @@ public int getCardinality() {
 
   @Override
   public int getLengthOfShortestElement() {
-    // Length of longest string
-    int minStringLength = Integer.MAX_VALUE;
-
-    // If this column is a string/bytes column, iterate over the dictionary to find the maximum length
-    FieldSpec.DataType dataType = _dataSource.getDataSourceMetadata().getDataType();
-    int length = _dictionary.length();
-
-    if (dataType.equals(FieldSpec.DataType.STRING)) {
-      for (int i = 0; i < length; i++) {
-        minStringLength = Math.min(_dictionary.getStringValue(i).length(), minStringLength);
-      }
-    } else if (dataType.equals(FieldSpec.DataType.BYTES)) {
-      for (int i = 0; i < length; i++) {
-        minStringLength = Math.min(_dictionary.getBytesValue(i).length, minStringLength);
-      }
-    }
-
-    return minStringLength;
+    collectElementLengthIfNeeded();
+    return _minElementLength;
   }
 
   @Override
   public int getLengthOfLargestElement() {
-    // Length of longest string
-    int maximumStringLength = 0;
+    collectElementLengthIfNeeded();
+    return _maxElementLength;
+  }
 
-    // If this column is a string/bytes column, iterate over the dictionary to find the maximum length
-    FieldSpec.DataType dataType = _dataSource.getDataSourceMetadata().getDataType();
-    int length = _dictionary.length();
+  private void collectElementLengthIfNeeded() {
+    if (_minElementLength >= 0) {
```

Comment:
is this check done so we don't calculate the element length more than once for a given dictionary ?

Reply:
Yes. Currently we scan the dictionary multiple times because we call `getLengthOfShortestElement()` and `getLengthOfLargestElement()` multiple times. This avoids that.

Reply:
Ah ok. I was going to suggest using an explicit flag to indicate that we have already collected the column stats. It won't be obvious for someone new reading code. anyway, we can fix another time.